Backend changes to enable third party game creation

blackshadowshade opened the pull request:
Starts to address #1784, but should leave user-visible behaviour untouched.

Note that it should now be possible to create third-party games by directly addressing the responder, e.g., via the Python client.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1219/ (if it passes)

blackshadowshade commented on src/engine/BMInterfaceGame.php:
```diff
@@ -49,15 +48,29 @@ public function create_game(
         try {
-            $gameId = $this->insert_new_game($playerIdArray, $maxWins, $description, $previousGameId);
+            if (!isset($currentPlayerId)) {
+                $currentPlayerId = $playerIdArray[0];
```
BTW, this line is here to maintain current behaviour with unit tests and external game creation via Python, etc.

I'm thinking that it might actually be better to throw an exception if the $currentPlayerId is NULL, but that may break stuff with your current model of game creation. Let me know what you think.

cgolubi1 replied:
Ten seconds of thinking: we think that throwing an exception should be fine; no one should be able to create games without a current player right now because it's in the authenticated part of BMInterface.php.

irilyth commented:
I've been wasting an inordinate amount of time lately pestering people to create UBFC and Mad Robin games, so if this would be easy to include in the December merge to dev, I'd love to try it out there. Especially if the existing python-client code can use it; I wasn't sure if the implication was that it would be with changes, or that it should work just as-is.

danlangford commented:
Fantastic, can't wait to try this
On Wed, Nov 30, 2016 at 9:26 AM irilyth <notifications@github.com> wrote:

> I've been wasting an inordinate amount of time lately pestering people to
> create UBFC and Mad Robin games, so if this would be easy to include in the
> December merge to dev, I'd love to try it out there. Especially if the
> existing python-client code can use it; I wasn't sure if the implication
> was that it would be with changes, or that it should work just as-is.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/buttonmen-dev/buttonmen/pull/2087#issuecomment-263919923>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAsB1HEXzTqL2vQ-iMCiuui7WvBg8Ibaks5rDaOzgaJpZM4JVFcj>
> .
>

blackshadowshade commented:
So, the existing python-client code should be able to use it, especially if it's running as if it were logged in as a user. Once this is merged, I'd suggest setting up a new account, maybe something like UBFC-Tourney, and then creating games logged in as that user.

blackshadowshade commented:
I've added the exception for when $currentPlayerId is unspecified. I've also rebased this into one commit.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1293/ (if it passes)

blackshadowshade commented:
I've checked the status of this, and it seems to be okay for testing.

cgolubi1 commented:
I don't think there's any point to standing up a dev site for this --- what it needs is replay testing, so i need to stop futzing with other pulls for long enough to get the wheels back on the replay engine after the last of the RandomAI turbo bugs, and then i can start testing this.  This weekend for sure (i hope)!

cgolubi1 commented:
I've created a replay site for this pull.

cgolubi1 commented:
I've replayed 1300 games against this, which, given the shallowness of the changes, seems like plenty (i.e. they're not going to trigger on weird mid-game skill combinations or anything).  So i'm going to tear the replay site down and declare victory (for my future reference, it's `ip-172-31-9-63`).

I think i should stand up a dev site to see if anyone sees anything weird, and because the API should allow third-party game creation, so e.g. irilyth should be able to try that out.

cgolubi1 commented on src/engine/BMInterfaceGame.php:
```diff
@@ -49,23 +48,43 @@ public function create_game(
         try {
-            $gameId = $this->insert_new_game($playerIdArray, $maxWins, $description, $previousGameId);
+            if (!isset($currentPlayerId)) {
+                if ($this->isTest) {
```
Heads up: putting test flags in non-test code makes me cringe.  I'm not going to veto it in the spirit of keeping things moving, but ideally there'd be a better way to do this.

blackshadowshade replied:
I'm open to suggestions.

cgolubi1 replied:
Hmm.  Since `$currentPlayerId` is an argument to `create_game()`, why can't test code set that argument to be the same as `$playerIdArray[0]` and pass it, rather than asking non-test code to do that?

blackshadowshade replied:
So, change all the multiple calls to create_game() in the test code? I believe that there are many of these.

cgolubi1 replied:
I hear you, but if you're changing the interface, you have to change the tests.

One thing i do a lot in responderTest is to create a wrapper around an API call, if there's a common way i always want to call that API.  Then i only have to implement to API changes in one place.

blackshadowshade replied:
Or create a different function that is used in the test code to replace the calls to create_game()?

cgolubi1 replied:
Yeah, that latter thing sounds like the wrapper idea, and probably makes your life easier here, because you can literally find/replace the calls to create_game with test_create_game_wrapper or whatever, and have that function set `$currentPlayerId` if it isn't set.  That may or may not be what you want for all invocations, so putting the change in the test code gives you the options.

For instance, suppose you want to make a test that demonstrates that `create_game()` fails in a clean way if `$currentPlayerId` isn't set?  Right now, you can't write that test.  But if you had a wrapper, you could simply not use the wrapper for that particular test.

blackshadowshade replied:
Okay, sounds reasonable. I'll take a look at doing that once preliminary testing is done.

blackshadowshade commented:
So, I've removed that test flag from the live code, added a new test function, and updated the unit tests to match.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1339/ (if it passes)

irilyth commented:
In the course of testing this, I found two bugs with the current game-creation API, #2218 and #2219; meanwhile, this change introduces a new bug, whereby I can create a game that specifies only the player's button, but not the player, opponent, or opponent's button. This is a regression from master, which currently says "ValueError: Failed to call createGame, got: Game create failed because you must be the first player."

I think that until/unless we want to support open games with neither player specified, the API should require you to set 'player' at a minimum -- it doesn't have to be yourself any more, but it has to be *someone*.

blackshadowshade commented:
I've addressed the regression by requiring that both the first player and the first button need to be specified.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1340/ (if it passes)

cgolubi1 commented:
The changes here look good - i'll go ahead and load this onto the dev site.

cgolubi1 commented:
irilyth: this is on your plate now to confirm that it looks good and afayk it's ready to merge.

irilyth commented:
Could you fix the DB so that http://blip.dev.buttonweavers.com/ui/open_games.html works? If replacing it with a clean DB is the easy way to do that, that's fine -- I just want to get it working again so I can make sure that none of my testing breaks it again.

blackshadowshade commented:
I have run the following SQL command for the blip instance:

    update game_player_map set player_id=1 where player_id is NULL and position=0;

I hope that gets you your overview page back.

blackshadowshade commented:
Oh, sorry, didn't read carefully enough to realise that you were referring to the open games.

I've now run

    update game_player_map set button_id=1 where position=0 and button_id is null;

Let me know if that fixes the open games page.

irilyth commented:
http://blip.dev.buttonweavers.com/ui/open_games.html looks all right now, yay. I'll do some more testing soon!

irilyth commented:
Ok, so here's some tests I ran.

# Games where the creator is the player

These tests are to verify that things work correctly when I (irilyth) create games where I'm the player.

## No buttons or players specified

    +$ ./game-create --site blip
    Traceback (most recent call last):
      File "./game-create", line 81, in <module>
	retval = bmconn.wrap_create_game(args.player_button, args.opponent_button, args.player, args.opponent, args.description)
      File "lib/bmutils.py", line 100, in wrap_create_game
	raise ValueError, "Failed to call createGame, got: " + retval.message
    ValueError: Failed to call createGame, got: The first button needs to be set.

That returns an error, as it should, because there isn't a way for player to pick a button after the game is created.

## Specifying only a player button, no players

    +$ ./game-create --site blip --player-button Angel
    Game 34 created

That created an open game, and automatically set me as the player, even though I hadn't specified myself; that seems right.

## Specifying myself as a player, and only a player button

    +$ ./game-create --site blip --player-button Angel --player irilyth
    Game 35 created

That created an open game that was identical except for game IDs and timestamps:

    +$ ./game-data --site blip 34 > ~/game-34.yaml
    +$ ./game-data --site blip 35 > ~/game-35.yaml

    +$ diff -U0 ~/game-34.yaml ~/game-35.yaml 
    --- /home/jbs/game-34.yaml      2017-06-30 16:27:46.989614333 -0400
    +++ /home/jbs/game-35.yaml      2017-06-30 16:27:50.297706838 -0400
    @@ -10 +10 @@
    -gameId: 34
    +gameId: 35
    @@ -54 +54 @@
    -  lastActionTime: 1498854243
    +  lastActionTime: 1498854426
    @@ -73 +73 @@
    -timestamp: 1498854243
    +timestamp: 1498854426

That seems right. I didn't bother to test all the other no-player-specified options with myself specified as a player, I'm pretty confident they'll all work as expected.

## Specifying a player button, and an opponent, but no opponent button

    +$ ./game-create --site blip --player-button Angel --opponent devious
    Game 36 created

That seems to have created a game, much as requested:

    +$ ./game-data --site blip --format json 36 | jp 'playerDataArray[*].{Name: playerName, Button: button.name}' | j2y
    - Button: Angel
      Name: irilyth
    - Button: ''
      Name: devious

It doesn't seem to show up anywhere in the UI (not on my games list, nor on the open games list), and http://blip.dev.buttonweavers.com/ui/game.html?game=36 just says "The game hasn't started yet."

Ah, this looks like a manifestation of #2219, which I'd thought was fixed, but apparently is not. Ok.

## Specifying a player button, and an opponent button, but no opponent

    +$ ./game-create --site blip --player-button Angel --opponent-button Bluff
    Game 37 created

That created an open game where both my button and the opponent's button were specified, which seems right.

## Specifying a player button, an opponent, and an oppponent button

    +$ ./game-create --site blip --player-button Angel --opponent devious --opponent-button Bluff
    Game 38 created

That created a regular game, waiting for devious to accept it.

I created a game just like that via the web UI (game 39), and they're identical except for game IDs and timestamps:

    +$ ./game-data --site blip 38 > ~/game-38.yaml
    +$ ./game-data --site blip 39 > ~/game-39.yaml

    +$ diff -U0 ~/game-38.yaml ~/game-39.yaml 
    --- /home/jbs/game-38.yaml	2017-06-30 23:17:24.321989528 -0400
    +++ /home/jbs/game-39.yaml	2017-06-30 23:17:27.970094109 -0400
    @@ -10 +10 @@
    -gameId: 38
    +gameId: 39
    @@ -40 +40 @@
    -  lastActionTime: 1498878254
    +  lastActionTime: 1498878309
    @@ -63 +63 @@
    -  lastActionTime: 1498878254
    +  lastActionTime: 1498878309
    @@ -83 +83 @@
    -timestamp: 1498878254
    +timestamp: 1498878309

That seems good.

# Games where the creator is someone other than the player

These tests are to verify that things work correctly when I (irilyth) create games where someone else (iriclone) is the player.

## Specifying a different player but no player button

    +$ ./game-create --site blip --player iriclone
    Traceback (most recent call last):
      File "./game-create", line 81, in <module>
	retval = bmconn.wrap_create_game(args.player_button, args.opponent_button, args.player, args.opponent, args.description)
      File "lib/bmutils.py", line 100, in wrap_create_game
	raise ValueError, "Failed to call createGame, got: " + retval.message
    ValueError: Failed to call createGame, got: The first button needs to be set.

That returns an error, as it should, because there isn't a way for player to pick a button after the game is created.

## Specifying a different player, and only a player button

    +$ ./game-create --site blip --player iriclone --player-button Angel
    Game 40 created

That created an open game, with iriclone as the player. It's identical to the one I created with myself (implicitly) as the player, except for game IDs, timestamps, and player-specific stuff:

    +$ ./game-data --site blip 40 > ~/game-40.yaml

    +$ diff -U0 ~/game-34.yaml ~/game-40.yaml
    --- /home/jbs/game-34.yaml	2017-06-30 16:27:46.989614333 -0400
    +++ /home/jbs/game-40.yaml	2017-06-30 23:18:22.107647643 -0400
    @@ -2 +2 @@
    -currentPlayerIdx: 0
    +currentPlayerIdx: false
    @@ -10 +10 @@
    -gameId: 34
    +gameId: 40
    @@ -35 +35 @@
    -  playerColor: '#ddffdd'
    +  playerColor: '#dddddd'
    @@ -54 +54 @@
    -  lastActionTime: 1498854243
    +  lastActionTime: 1498878731
    @@ -57,3 +57,3 @@
    -  playerColor: '#dd99dd'
    -  playerId: 4
    -  playerName: irilyth
    +  playerColor: '#cccccc'
    +  playerId: 5
    +  playerName: iriclone
    @@ -73 +73 @@
    -timestamp: 1498854243
    +timestamp: 1498878731

That all seems good.

## Specifying a different player, a player button, and an opponent, but no opponent button

    +$ ./game-create --site blip --player iriclone --player-button Angel --opponent devious
    Game 41 created

That seems to have created a game, much as requested:

    +$ ./game-data --site blip --format json 41 | jp 'playerDataArray[*].{Name: playerName, Button: button.name}' | j2y
    - Button: Angel
      Name: iriclone
    - Button: ''
      Name: devious

It doesn't seem to show up anywhere in the UI, and http://blip.dev.buttonweavers.com/ui/game.html?game=41 just says "The game hasn't started yet.", so another manifestation of #2219.

## Specifying a different player, a player button, and an opponent button, but no opponent

    +$ ./game-create --site blip --player iriclone --player-button Angel --opponent-button Bluff
    Game 42 created

That created an open game where both iriclone's button and the opponent's button were specified, which seems right.

## Specifying a different player, a player button, an opponent, and an oppponent button

    +$ ./game-create --site blip --player iriclone --player-button Angel --opponent devious --opponent-button Bluff
    Game 43 created

That created a regular game, waiting for devious to accept it.

I created a game just like that via the web UI (game 44), and they're identical except for game IDs and timestamps:

    +$ ./game-data --site blip 43 > ~/game-43.yaml
    +$ ./game-data --site blip 44 > ~/game-44.yaml

    +$ diff -U0 ~/game-43.yaml ~/game-44.yaml 
    --- /home/jbs/game-43.yaml	2017-06-30 23:24:25.210142907 -0400
    +++ /home/jbs/game-44.yaml	2017-06-30 23:24:28.434236772 -0400
    @@ -10 +10 @@
    -gameId: 43
    +gameId: 44
    @@ -40 +40 @@
    -  lastActionTime: 1498879340
    +  lastActionTime: 1498879442
    @@ -63 +63 @@
    -  lastActionTime: 1498879340
    +  lastActionTime: 1498879442
    @@ -83 +83 @@
    -timestamp: 1498879339
    +timestamp: 1498879442

So that seems good: A game between iriclone playing Angel and devious playing Bluff looks the same whether created by iriclone via the web UI or by irilyth via the API.

## Specifying the creator as the opponent

    +$ ./game-create --site blip --player iriclone --player-button Angel --opponent irilyth --opponent-button Bluff
    Game 45 created

    +$ ./game-create --site blip --player irilyth --player-button Angel --opponent iriclone --opponent-button Bluff
    Game 46 created

Those are different in the expected ways, 'player' and 'opponent' are reversed (plus game IDs and timestamps are different):

    +$ ./game-data --site blip 45 > ~/game-45.yaml
    +$ ./game-data --site blip 46 > ~/game-46.yaml

    +$ diff -U0 ~/game-45.yaml ~/game-46.yaml
    --- /home/jbs/game-45.yaml	2017-06-30 23:41:49.970234107 -0400
    +++ /home/jbs/game-46.yaml	2017-06-30 23:41:57.042470825 -0400
    @@ -2 +2 @@
    -currentPlayerIdx: 1
    +currentPlayerIdx: 0
    @@ -10 +10 @@
    -gameId: 45
    +gameId: 46
    @@ -31 +31 @@
    -opponent: &id001
    +opponent: &id002
    @@ -33,7 +33 @@
    -  - description: Shadow 4-sided die
    -    properties: []
    -    recipe: s(4)
    -    sides: 4
    -    skills: [Shadow]
    -    value: null
    -  - description: 6-sided die
    +  - description: Poison Shadow 6-sided die
    @@ -41 +35 @@
    -    recipe: (6)
    +    recipe: ps(6)
    @@ -43 +37 @@
    -    skills: []
    +    skills: [Poison, Shadow]
    @@ -45 +39 @@
    -  - description: Shadow 12-sided die
    +  - description: Poison Shadow 12-sided die
    @@ -47 +41 @@
    -    recipe: s(12)
    +    recipe: ps(12)
    @@ -49 +43 @@
    -    skills: [Shadow]
    +    skills: [Poison, Shadow]
    @@ -51 +45 @@
    -  - description: 12-sided die
    +  - description: 16-sided die
    @@ -53,2 +47,8 @@
    -    recipe: (12)
    -    sides: 12
    +    recipe: (16)
    +    sides: 16
    +    skills: []
    +    value: null
    +  - description: 20-sided die
    +    properties: []
    +    recipe: (20)
    +    sides: 20
    @@ -63,2 +63,2 @@
    -  button: {artFilename: angel.png, name: Angel, originalRecipe: s(4) (6) s(12) (12)
    -      (X), recipe: s(4) (6) s(12) (12) (X)}
    +  button: {artFilename: bluff.png, name: Bluff, originalRecipe: ps(6) ps(12) (16)
    +      (20) (X), recipe: ps(6) ps(12) (16) (20) (X)}
    @@ -70 +70 @@
    -  lastActionTime: 1498880430
    +  lastActionTime: 1498880479
    @@ -84 +84 @@
    -player: &id002
    +player: &id001
    @@ -86 +86 @@
    -  - description: Poison Shadow 6-sided die
    +  - description: Shadow 4-sided die
    @@ -88,3 +88,3 @@
    -    recipe: ps(6)
    -    sides: 6
    -    skills: [Poison, Shadow]
    +    recipe: s(4)
    +    sides: 4
    +    skills: [Shadow]
    @@ -92 +92 @@
    -  - description: Poison Shadow 12-sided die
    +  - description: 6-sided die
    @@ -94,3 +94,3 @@
    -    recipe: ps(12)
    -    sides: 12
    -    skills: [Poison, Shadow]
    +    recipe: (6)
    +    sides: 6
    +    skills: []
    @@ -98 +98 @@
    -  - description: 16-sided die
    +  - description: Shadow 12-sided die
    @@ -100,3 +100,3 @@
    -    recipe: (16)
    -    sides: 16
    -    skills: []
    +    recipe: s(12)
    +    sides: 12
    +    skills: [Shadow]
    @@ -104 +104 @@
    -  - description: 20-sided die
    +  - description: 12-sided die
    @@ -106,2 +106,2 @@
    -    recipe: (20)
    -    sides: 20
    +    recipe: (12)
    +    sides: 12
    @@ -116,2 +116,2 @@
    -  button: {artFilename: bluff.png, name: Bluff, originalRecipe: ps(6) ps(12) (16)
    -      (20) (X), recipe: ps(6) ps(12) (16) (20) (X)}
    +  button: {artFilename: angel.png, name: Angel, originalRecipe: s(4) (6) s(12) (12)
    +      (X), recipe: s(4) (6) s(12) (12) (X)}
    @@ -123 +123 @@
    -  lastActionTime: 1498880430
    +  lastActionTime: 1498880479
    @@ -143 +143 @@
    -timestamp: 1498880430
    +timestamp: 1498880479

Hmm, but the one where iriclone is the player and irilyth is the opponent (game 45) isn't identical to the one from before where iriclone is the player and devious is the opponent (game 43)... Ah, because iriclone has auto-accept on, and devious (presumably) doesn't. I turned it off for iriclone, and created another game (game 47) just like 45:

    +$ ./game-create --site blip --player iriclone --player-button Angel --opponent irilyth --opponent-button Bluff
    Game 47 created

    +$ ./game-data --site blip 47 > ~/game-47.yaml

Those are closer to identical, except for game IDs, timestamps, and player-specific stuff:

    +$ diff -U0 ~/game-47.yaml ~/game-43.yaml
    --- /home/jbs/game-47.yaml	2017-06-30 23:51:36.474259915 -0400
    +++ /home/jbs/game-43.yaml	2017-06-30 23:24:25.210142907 -0400
    @@ -2 +2 @@
    -currentPlayerIdx: 1
    +currentPlayerIdx: false
    @@ -10 +10 @@
    -gameId: 47
    +gameId: 43
    @@ -31 +31 @@
    -opponent: &id001
    +opponent: &id002
    @@ -33,2 +33,2 @@
    -  button: {artFilename: angel.png, name: Angel, originalRecipe: null, recipe: s(4)
    -      (6) s(12) (12) (X)}
    +  button: {artFilename: bluff.png, name: Bluff, originalRecipe: null, recipe: ps(6)
    +      ps(12) (16) (20) (X)}
    @@ -40 +40 @@
    -  lastActionTime: 1498881086
    +  lastActionTime: 1498879340
    @@ -43,3 +43,3 @@
    -  playerColor: '#ddffdd'
    -  playerId: 5
    -  playerName: iriclone
    +  playerColor: '#dddddd'
    +  playerId: 6
    +  playerName: devious
    @@ -54 +54 @@
    -player: &id002
    +player: &id001
    @@ -56,2 +56,2 @@
    -  button: {artFilename: bluff.png, name: Bluff, originalRecipe: null, recipe: ps(6)
    -      ps(12) (16) (20) (X)}
    +  button: {artFilename: angel.png, name: Angel, originalRecipe: null, recipe: s(4)
    +      (6) s(12) (12) (X)}
    @@ -63 +63 @@
    -  lastActionTime: 1498881086
    +  lastActionTime: 1498879340
    @@ -66,3 +66,3 @@
    -  playerColor: '#dd99dd'
    -  playerId: 4
    -  playerName: irilyth
    +  playerColor: '#cccccc'
    +  playerId: 5
    +  playerName: iriclone
    @@ -83 +83 @@
    -timestamp: 1498881086
    +timestamp: 1498879339

The one other thing is that iriclone is player in the iriclone vs devious game, but opponent in the iriclone vs irilyth game, despite being --player on both commands:

    +$ ./game-create --site blip --player iriclone --player-button Angel --opponent devious --opponent-button Bluff
    Game 43 created

    +$ ./game-create --site blip --player iriclone --player-button Angel --opponent irilyth --opponent-button Bluff
    Game 47 created

So I guess if the creator is specified as opponent, the game is created with them as player? That's probably fine, even if it wasn't what I'd expected.

# Conclusions

So, two potential issues:

- #2219 is still open, which means that if someone creates a game that specifies an opponent but no opponent button, an oprhaned game gets created. This is true in the current code too, so probably not a blocker for this, and it's fine if we don't want to fix that until this is released.
- The thing whereby the if the creator is specified as opponent, the game is created with them as player, not as opponent. That's probably fine but I wanted to check.

Hmm, one other thing: I haven't tested the change from https://github.com/buttonmen-dev/buttonmen/pull/2087/commits/09ea4b9554a915bfa20927a216f558c02e5f9f58. The current Python API client doesn't require (and doesn't allow) you to specify player, and when none is specified, player is the creator. My modifications to bmapi.py allow you to specify player, but don't require it, and default to creator if it's not specified, to retain the current behavior... But that means that even if I don't specify player, I'm getting creator by default, and thus not testing the bug that 09ea4b9 fixed.

Well, I can test that by removing that default temporarily; lemme send this, and then I'll test that.

irilyth commented:
> I can create a game that specifies only the player's button, but not the player, opponent, or opponent's button. 

That was fixed in https://github.com/buttonmen-dev/buttonmen/pull/2087/commits/09ea4b9554a915bfa20927a216f558c02e5f9f58, and I tested it by temporarily removing the code that I'd added to bmapi.py, to use creator as player if player isn't specified:

    +    if player == None or player == '':
    +      player = self.username

I commented that out, and tried the test that broke things before:

    +$ ./game-create --site blip --player-button Angel
    Traceback (most recent call last):
      File "./game-create", line 81, in <module>
	retval = bmconn.wrap_create_game(args.player_button, args.opponent_button, args.player, args.opponent, args.description)
      File "lib/bmutils.py", line 100, in wrap_create_game
	raise ValueError, "Failed to call createGame, got: " + retval.message
    ValueError: Failed to call createGame, got: Game create failed because the first player was not specified.

So, that seems right; yay.

Check out the two comments in "Conclusions" in the previous update, and if those don't seem like issues, I think this is ready to ship!

cgolubi1 commented:
> The thing whereby the if the creator is specified as opponent, the game is created with them as player, not as opponent. That's probably fine but I wanted to check.

Do we understand why this happens?  It seems like it could be mildly undesirable from a code complexity standpoint, and we are storing the `creatorId` separately in the database now.  So i'm curious if we understand why it's happening, and then we can weigh in on whether it's good.

Another thing i'm not sure we've tested: have we tried loading a database containing existing games under this new code?  I know the `creator_id` field already existed in the DB before this change, but we should make sure games created under previous code remain loadable.  I think this has to be a manual test, so i can do that under virtualbox if we think we're otherwise ready to go.

irilyth commented:
> Another thing i'm not sure we've tested: have we tried loading a database containing existing games under this new code? I know the creator_id field already existed in the DB before this change, but we should make sure games created under previous code remain loadable. I think this has to be a manual test, so i can do that under virtualbox if we think we're otherwise ready to go.

Aren't all of the dev sites "databases with existing games" that come up under the new code?

blackshadowshade commented:
@irilyth, in game 45 where you specified yourself as the game creator and the second player, I see the following in the header:

    Game #45  •  iriclone (Angel) vs. irilyth (Bluff)  •  Round #1

The allocation of 'player' and 'opponent' is dependent on who is loading the game, and when you are one of the players, then 'player' is assigned to you.

Also, when I run the following SQL query on the database

    select m.player_id, m.position, p.name_ingame 
    from game_player_map as m 
    left join player as p 
    on p.id=m.player_id 
    where game_id=45;

I get the following:

    +-----------+----------+-------------+
    | player_id | position | name_ingame |
    +-----------+----------+-------------+
    |         4 |        1 | irilyth     |
    |         5 |        0 | iriclone    |
    +-----------+----------+-------------+
    2 rows in set (0.00 sec)

That looks right to me.

cgolubi1 commented:
Ah, yeah, now i understand: i was thinking irilyth was talking about position 0 vs. position 1, but of course the player/opponent representation is always going to depend on who's loading the game, so i don't see a problem there.

cgolubi1 commented:
> Aren't all of the dev sites "databases with existing games" that come up under the new code?

Yeah, you're right - i loaded the first several games on the blip dev site (which predated this code change), and they looked fine.

cgolubi1 commented:
Okay, i think we're at "ship it".  Shadowshade, rebase this down to one commit, and let's merge it.

blackshadowshade commented:
Rebase done.

New Jenkins: http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1347/ (if it passes)

cgolubi1 commented:
Awesome, thanks!  Merging.
